test/composableStableExit.integration.test.ts

```diff
@@ -0,0 +1,336 @@
+    const bptIn = TokenAmount.fromHumanAmount(bptToken, '1');
+
+    const exitInput: ProportionalExitInput = {
+      chainId,
+      rpcUrl,
+      bptIn,
+      kind: ExitKind.PROPORTIONAL,
+    };
+    const { queryResult, maxBptIn, minAmountsOut } = await doTransaction({
+      ...txInput,
+      exitInput,
+    });
+
+    // Query should use correct BPT amount
+    expect(queryResult.bptIn.amount).to.eq(bptIn.amount);
+
+    // We expect all assets to have a value for exit
+    expect(queryResult.tokenOutIndex).to.be.undefined;
+    queryResult.amountsOut.filter((_, index) => index!==bptIndex).forEach((a) => {
+      expect(a.amount > BigInt(0)).to.be.true;
+    });
+
+    // Confirm slippage - only to amounts out not bpt in
+    const expectedMinAmountsOut = queryResult.amountsOut.map((a) =>
+      slippage.removeFrom(a.amount),
+    );
+    expect(expectedMinAmountsOut).to.deep.eq(minAmountsOut);
+    expect(maxBptIn).to.eq(bptIn.amount);
+  });
+
+  test('unbalanced exit', async () => {
+    const bptIndex = txInput.poolInput.tokens.findIndex((t) => t.address === txInput.poolInput.address);
+    const { poolInput, slippage } = txInput;
+
+    const poolTokens = poolInput.tokens
+      .filter((_, index) => index!==bptIndex)
+      .map(
+      (t) => new Token(chainId, t.address, t.decimals),
+    );
+    const amountsOut = poolTokens.map((t) =>
+      TokenAmount.fromHumanAmount(t, '80'),
+    );
+
+    const exitInput: UnbalancedExitInput = {
+      chainId,
+      rpcUrl,
+      amountsOut,
+      kind: ExitKind.UNBALANCED,
+    };
+    const { queryResult, maxBptIn, minAmountsOut } = await doTransaction({
+      ...txInput,
+      exitInput,
+    });
+
+    // We expect a BPT input amount > 0
+    expect(queryResult.bptIn.amount > BigInt(0)).to.be.true;
+
+    // We expect assets to have same amount out as user defined
+    expect(queryResult.tokenOutIndex).to.be.undefined;
+    queryResult.amountsOut.filter((_, index) => index!==bptIndex).forEach((a, i) => {
+      expect(a.amount).to.eq(amountsOut[i].amount);
+    });
+
+    // Confirm slippage - only to bpt in, not amounts out
+    const expectedMinAmountsOut = amountsOut.map((a) => a.amount);
+    expect(expectedMinAmountsOut).to.deep.eq(minAmountsOut.filter((_, index) => index!==bptIndex));
+    const expectedMaxBptIn = slippage.applyTo(queryResult.bptIn.amount);
+    expect(expectedMaxBptIn).to.deep.eq(maxBptIn);
+  });
+
+  test('exit with native asset', async () => {
+    const bptIndex = txInput.poolInput.tokens.findIndex((t) => t.address === txInput.poolInput.address);
+
+    const { slippage } = txInput;
+    const bptIn = TokenAmount.fromHumanAmount(bptToken, '1');
+
+    const exitInput: ProportionalExitInput = {
+      chainId,
+      rpcUrl,
+      bptIn,
+      kind: ExitKind.PROPORTIONAL,
+      exitWithNativeAsset: true,
+    };
+
+    // Note - checking native balance
+    const { queryResult, maxBptIn, minAmountsOut } = await doTransaction({
+      ...txInput,
```

Comment:
Similar to join lets reuse any similar helper functions across tests.

Reply:
But the composable doTransaction has a different expectedAmounts than the weighted, maybe I could use the same function, but adapt for each pool type?

Reply:
I think there's a lot of similar functionality that could be reused?